lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -578,6 +587,61 @@ findMemberReference(Expr *expr) {
   return None;
 }
 
+static bool isAsyncCall(const ApplyExpr *call) {
+  // Effectively the same as doing a
+  // `dyn_cast_or_null<FunctionType>(call->getFn()->getType())`, check the
+  // result of that and then checking `isAsync` if it's defined.
+  Type funcTypeType = call->getFn()->getType();
+  if (!funcTypeType)
+    return false;
+  FunctionType *funcType = funcTypeType->getAs<FunctionType>();
+  return funcType && funcType->isAsync();
+}
+
+static llvm::SmallVector<InOutArgInfo, 3> extractAsyncInoutArgs(const ApplyExpr *call) {
+  // Is the function we're calling async? If not, it doesn't matter.
+  // Is the function we're calling a mutating method? (implicit inout of 'self')
+  // gather the inout parameters recursively
+
+  // 1.
+  // Is it async?
+  if (!isAsyncCall(call))
+    return {};
+
+  llvm::SmallVector<InOutArgInfo, 3> outputs;
+  // Collect all of the inout arguments from the arguments of an expression.
+  // This does not collect things that have been curried in
+  // isPartialApply should be true if the parameter is part of the partial
+  // application.
+  auto extractInoutArgs = [&outputs](const ApplyExpr *call,
+                                     bool isPartialApply=false) {
+    if (ParenExpr *arg = dyn_cast<ParenExpr>(call->getArg())) {
```

Comment:
I worry about code like `callee(((arg)))` here. You might want to skip this ParenExpr dyn_cast and use `callee->getArg()->getSemanticsProvidingExpr()` to look through arbitrary paren nesting. That should then allow you to simplify this if-else block a bit.